src/gradedunitrange.jl

```diff
@@ -0,0 +1,278 @@
+using BlockArrays:
+  BlockArrays,
+  AbstractBlockVector,
+  AbstractBlockedUnitRange,
+  Block,
+  BlockIndex,
+  BlockIndexRange,
+  BlockRange,
+  BlockSlice,
+  BlockVector,
+  BlockedOneTo,
+  block,
+  blockedrange,
+  blocklasts,
+  blocklengths,
+  blocks,
+  blockindex,
+  combine_blockaxes,
+  findblock,
+  mortar
+using BlockSparseArrays:
+  BlockSparseArrays,
+  blockedunitrange_findblock,
+  blockedunitrange_findblockindex,
+  blockedunitrange_getindices
+using Compat: allequal
+
+abstract type AbstractGradedUnitRange{T,BlockLasts} <:
+              AbstractBlockedUnitRange{T,BlockLasts} end
+
+struct GradedUnitRange{T,SUR<:SectorOneTo{T},BR<:AbstractUnitRange{T},BlockLasts} <:
+       AbstractGradedUnitRange{T,BlockLasts}
+  sector_axes::Vector{SUR}
+  full_range::BR
+
+  function GradedUnitRange{T,SUR,BR,BlockLasts}(
+    sector_axes::AbstractVector{SUR}, full_range::AbstractUnitRange{T}
+  ) where {T,SUR,BR,BlockLasts}
+    length.(sector_axes) == blocklengths(full_range) ||
+      throw(ArgumentError("sectors and range are not compatible"))
+    allequal(isdual.(sector_axes)) ||
+      throw(ArgumentError("all blocks must have same duality"))
+    typeof(blocklasts(full_range)) == BlockLasts ||
+      throw(TypeError(:BlockLasts, "", blocklasts(full_range)))
+    return new{T,SUR,BR,BlockLasts}(sector_axes, full_range)
+  end
+end
+
+const GradedOneTo{T,SUR,BR,BlockLasts} =
+  GradedUnitRange{T,SUR,BR,BlockLasts} where {BR<:BlockedOneTo}
+
+function GradedUnitRange(sector_axes::AbstractVector, full_range::AbstractUnitRange)
+  return GradedUnitRange{
+    eltype(full_range),eltype(sector_axes),typeof(full_range),typeof(blocklasts(full_range))
+  }(
+    sector_axes, full_range
+  )
+end
+
+# Accessors
+sector_axes(g::GradedUnitRange) = g.sector_axes
+unlabel_blocks(g::GradedUnitRange) = g.full_range  # TBD rename full_range?
```

Comment:
Partially as a reminder to myself: I think we decided to rename this to `ungrade`, which would be a generic function that can be used on any graded array (and does nothing on non-graded arrays).